### PR TITLE
Merge master to hlsl-2021

### LINKED ISSUE #3696: Rebase hlsl-2021 branch on top of master
It would be really nice to refresh the hlsl-2021 branch so they contain all the recent fixes that have gone into master, looks like it's a relatively straight forward rebase/merge operation.

### PR DESCRIPTION
**Pull request by hekota**
Fixes #3696

### CODE REVIEW

**Review by pow2clk — APPROVED** [review]
Yes, I have definitely reviewed every line of this change and confirmed that they match the changes submitted to the main branch character for character 😅

File: tools/clang/lib/CodeGen/CGDebugInfo.cpp [hunk 1]
```diff
@@ -466,8 +466,11 @@ llvm::DIType *CGDebugInfo::CreateType(const BuiltinType *BT) {
   case BuiltinType::OCLEvent:
     llvm_unreachable("No ObjC or OpenCL support");
   case BuiltinType::LitInt:
+    Encoding = llvm::dwarf::DW_ATE_signed;
+    break;
   case BuiltinType::LitFloat:
-    llvm_unreachable("Unsupported HLSL types");
+    Encoding = llvm::dwarf::DW_ATE_float;
+    break;
```
Comment (pow2clk) [inline tools/clang/lib/CodeGen/CGDebugInfo.cpp:473]: The fault really lies with me for adding these here in the first place, but a better place for these would be the HLSL-specific sections of cases below including Min16Int  and HalfFloat respectively. This `#else` block is meant for ObjC and OCL stuff.

File: tools/clang/test/HLSLFileCheck/hlsl/template/AddMulOps.hlsl [hunk 2]
```diff
@@ -1,5 +1,6 @@
 // RUN: %dxc -E main -T ps_6_0 -enable-templates %s | FileCheck %s
 // RUN: %dxc -E main -T ps_6_0 %s -enable-templates -DCHECK_DIAGNOSTICS | FileCheck %s -check-prefix=DIAG
+// RUN: %dxc -E main -T ps_6_0 -enable-templates %s /Zi
```
Comment (pow2clk) [inline tools/clang/test/HLSLFileCheck/hlsl/template/AddMulOps.hlsl:3]: No Filecheck, intended just to make sure it doesn't crash?

**Review by pow2clk — APPROVED** [review]
the changes are fine as-is, but I've suggested a slightly better organization.

File: tools/clang/lib/CodeGen/CGDebugInfo.cpp (hunk 1 above)
Comment (hekota) [reply tools/clang/lib/CodeGen/CGDebugInfo.cpp]: fixed

**Comment by tex3d** [comment]
> Yes, I have definitely reviewed every line of this change and confirmed that they match the changes submitted to the main branch character for character 😅

If you look at the merge commit, you can see the conflict resolution (changes introduced during merge in addition to automatic merge of incoming changes), though I'm not sure how you can do this in GitHub.  Here they are:
```
---------------------- include/dxc/Support/HLSLOptions.td ----------------------
index 20bbe0eeb,5679f63bb..22daf4878
@@@ -277,8 -277,12 +277,14 @@@ def enable_lifetime_markers : Flag<["-"
    HelpText<"Enable generation of lifetime markers">;
  def disable_lifetime_markers : Flag<["-", "/"], "disable-lifetime-markers">, Group<hlslcomp_Group>, Flags<[CoreOption, HelpHidden]>,
    HelpText<"Disable generation of lifetime markers where they would be otherwise (6.6+)">;
 +def enable_templates: Flag<["-", "/"], "enable-templates">, Group<hlslcomp_Group>, Flags<[CoreOption]>,
 +  HelpText<"Enable template support for HLSL.">;
+ def enable_payload_qualifiers : Flag<["-", "/"], "enable-payload-qualifiers">, Group<hlslcomp_Group>, Flags<[CoreOption, RewriteOption, DriverOption]>,
+   HelpText<"Enables support for payload access qualifiers for raytracing payloads in SM 6.6.">;
+ def disable_payload_qualifiers : Flag<["-", "/"], "disable-payload-qualifiers">, Group<hlslcomp_Group>, Flags<[CoreOption, RewriteOption, DriverOption]>,
+   HelpText<"Disables support for payload access qualifiers for raytracing payloads in SM 6.7.">;
+ def disable_exception_handling : Flag<["-", "/"], "disable-exception-handling">, Group<hlslcomp_Group>, Flags<[DriverOption, HelpHidden]>,
+   HelpText<"Disable dxc handling of exceptions">;
  
  // Used with API only
  def skip_serialization : Flag<["-", "/"], "skip-serialization">, Group<hlslcore_Group>, Flags<[CoreOption, HelpHidden]>,

---------------- tools/clang/include/clang/Basic/LangOptions.h ----------------
index 22dac3080,57ac9d1bf..b258c47b8
@@@ -158,7 -158,7 +158,8 @@@ public
    bool UseMinPrecision; // use min precision, not native precision.
    bool EnableDX9CompatMode;
    bool EnableFXCCompatMode;
 +  bool EnableTemplates;
+   bool EnablePayloadAccessQualifiers;
    // HLSL Change Ends
  
    bool SPIRV = false;  // SPIRV Change

```

Just straight-forward adjacency conflict resolutions, it would appear.

The rest of the differences are the already-verified incoming commits, and the additional commits in this PR.

**Review by tex3d — APPROVED** [review]
LGTM!

**Pull request merged**